File: ui-automation/selenium-javascript/tests/heroku/loginFunctionality.test.js
const { expect } = require('chai');
const BaseTest = require('../../utils/BaseTest');
const HerokuLoginPage = require('../../pages/HerokuLoginPage');

describe('Login Functionality', function() {
    let baseTest;
    let driver;
    let loginPage;

    beforeEach(async function() {
        baseTest = new BaseTest();
        driver = await baseTest.setUp();
        loginPage = new HerokuLoginPage(driver);
    });

    afterEach(async function() {
        await baseTest.tearDown();
    });

    it('should display an error for invalid credentials', async function() {
        // Navigate to login page
        await loginPage.navigateToLoginPage();

        // Enter invalid credentials and login
        await loginPage.login('invalidUser', 'invalidPass');

        // Wait for response
        await driver.sleep(2000);

        // Verify error message
        const flashMessage = await loginPage.getFlashMessage();
        expect(flashMessage).to.include('Your username is invalid!');
    });

    it('should successfully log in with valid credentials', async function() {
        // Navigate to login page
        await loginPage.navigateToLoginPage();

        // Enter valid credentials and login
        await loginPage.login('tomsmith', 'SuperSecretPassword!');

        // Wait for response
        await driver.sleep(2000);

        // Verify successful login
        const isOnSecurePage = await loginPage.isOnSecurePage();
        expect(isOnSecurePage).to.be.true;

        const flashMessage = await loginPage.getFlashMessage();
        expect(flashMessage).to.include('You logged into a secure area!');
    });

    it('should successfully log out', async function() {
        // Navigate to login page and log in first
        await loginPage.navigateToLoginPage();
        await loginPage.login('tomsmith', 'SuperSecretPassword!');

        // Wait for login to complete
        await driver.sleep(2000);

        // Log out
        await loginPage.clickLogoutButton();

        // Wait for logout to complete
        await driver.sleep(2000);

        // Verify successful logout
        const isOnLoginPage = await loginPage.isOnLoginPage();
        expect(isOnLoginPage).to.be.true;

        const flashMessage = await loginPage.getFlashMessage();
        expect(flashMessage).to.include('You logged out of the secure area!');
    });
});

File: ui-automation/selenium-java/src/test/java/tests/heroku/LoginFunctionalityTest.java
package tests.heroku;

import org.testng.annotations.Test;
import org.testng.Assert;
import pages.HerokuLoginPage;
import utils.BaseTest;

/**
 * Test class for Login Functionality
 * Mirrors the loginFunctionality.cy.js Cypress test
 */
public class LoginFunctionalityTest extends BaseTest {
    
    @Test(description = "Should display an error for invalid credentials")
    public void shouldDisplayErrorForInvalidCredentials() {
        HerokuLoginPage loginPage = new HerokuLoginPage();
        
        // Navigate to login page
        loginPage.navigateToLoginPage();
        
        // Enter invalid credentials
        loginPage.login("invalidUser", "invalidPass");
        
        // Verify error message
        Assert.assertTrue(loginPage.getFlashMessage().contains("Your username is invalid!"), 
            "Flash message should contain 'Your username is invalid!'");
    }
    
    @Test(description = "Should successfully log in with valid credentials")
    public void shouldSuccessfullyLogInWithValidCredentials() {
        HerokuLoginPage loginPage = new HerokuLoginPage();
        
        // Navigate to login page
        loginPage.navigateToLoginPage();
        
        // Enter valid credentials
        loginPage.login("tomsmith", "SuperSecretPassword!");
        
        // Verify successful login
        Assert.assertTrue(loginPage.isOnSecurePage(), 
            "Should be on secure page after successful login");
        Assert.assertTrue(loginPage.getFlashMessage().contains("You logged into a secure area!"), 
            "Flash message should contain 'You logged into a secure area!'");
    }
    
    @Test(description = "Should successfully log out")
    public void shouldSuccessfullyLogOut() {
        HerokuLoginPage loginPage = new HerokuLoginPage();
        
        // Navigate to login page and log in first
        loginPage.navigateToLoginPage();
        loginPage.login("tomsmith", "SuperSecretPassword!");
        
        // Log out
        loginPage.clickLogoutButton();
        
        // Verify successful logout
        Assert.assertTrue(loginPage.isOnLoginPage(), 
            "Should be back on login page after logout");
        Assert.assertTrue(loginPage.getFlashMessage().contains("You logged out of the secure area!"), 
            "Flash message should contain 'You logged out of the secure area!'");
    }
}
